time-tree.js

```diff
@@ -0,0 +1,88 @@
+'use strict'
+const archy = require('archy')
+
+function TimeTree () {
+  this.root = null
+  this.tableId = new Map()
+  this.tableLabel = new Map()
+}
+
+TimeTree.prototype.trackNode = function (node) {
+  this.tableId.set(node.id, node)
+  this.tableLabel.set(node.label, node)
+}
+
+TimeTree.prototype.untrackNode = function (node) {
+  this.tableId.delete(node.id)
+  this.tableLabel.delete(node.label)
+}
+
+TimeTree.prototype.getParent = function (parent) {
+  if (parent === null) {
+    return this.root
+  }
+
+  return this.tableLabel.get(parent)
+}
+
+TimeTree.prototype.getNode = function (nodeId) {
+  return this.tableId.get(nodeId)
+}
+
+TimeTree.prototype.add = function (parent, child, start) {
+  const isRoot = parent === null
+  if (isRoot) {
+    this.root = {
+      id: 'root',
+      label: child,
+      start,
+      nodes: []
+    }
+    this.trackNode(this.root)
+    return this.root.id
+  }
+
+  const parentNode = this.getParent(parent)
+  const nodeId = `${child}-${Math.random()}`
+  const childNode = {
+    id: nodeId,
+    parent,
+    start,
+    label: child,
+    nodes: []
+  }
+  parentNode.nodes.push(childNode)
+  this.trackNode(childNode)
+  return nodeId
+}
+
+TimeTree.prototype.start = function (parent, child, start) {
+  return this.add(parent, child, start || Date.now())
+}
+
+TimeTree.prototype.stop = function (nodeId, stop) {
+  const node = this.getNode(nodeId)
+  if (node) {
+    node.stop = stop || Date.now()
+    node.diff = (node.stop - node.start) || 0
+    this.untrackNode(node)
+  }
+}
+
+TimeTree.prototype.toJSON = function () {
+  return Object.assign({}, this.root)
+}
+
+TimeTree.prototype.prittyPrint = function () {
+  const decorateText = (node) => {
+    node.label = `${node.label} ${node.diff} ms`
+    if (node.nodes.length > 0) {
+      node.nodes = node.nodes.map(_ => decorateText(_))
+    }
+    return node
+  }
+  const out = decorateText(this.toJSON())
+  console.log(archy(out))
```

Comment:
I think that `prettyPrint` should return a string instead of print it.
It should also make it easier to test.

Reply:
Done